custom_components/pyscript/requirements.py

```diff
@@ -0,0 +1,223 @@
+from packaging.version import Version
+from pkg_resources import Requirement
+
+from homeassistant.loader import bind_hass
+from homeassistant.requirements import async_process_requirements
+
+from .const import (
+    ATTR_INSTALLED_VERSION,
+    ATTR_SOURCES,
+    ATTR_VERSION,
+    CONF_INSTALLED_PACKAGES,
+    DOMAIN,
+    LOGGER_PATH,
+    REQUIREMENTS_FILE,
+    REQUIREMENTS_PATHS,
+)
+
+if sys.version_info[:2] >= (3, 8):
+    from importlib.metadata import (  # pylint: disable=no-name-in-module,import-error
+        PackageNotFoundError,
+        version as installed_version,
+    )
+else:
+    from importlib_metadata import (  # pylint: disable=import-error
+        PackageNotFoundError,
+        version as installed_version,
+    )
+
+_LOGGER = logging.getLogger(LOGGER_PATH)
+
+
+def get_installed_version(pkg_name):
+    """Get installed version of package. Returns None if not found."""
+    try:
+        return installed_version(pkg_name)
+    except PackageNotFoundError:
+        return None
+
+
+@bind_hass
+def process_all_requirements(pyscript_folder, requirements_paths, requirements_file):
+    """
+    Load all lines from requirements_file located in requirements_paths.
+
+    Returns files and a list of packages, if any, that need to be installed.
+    """
+    all_requirements_to_process = {}
+    for root in requirements_paths:
+        for requirements_path in glob.glob(os.path.join(pyscript_folder, root, requirements_file)):
+            with open(requirements_path, "r") as requirements_fp:
+                all_requirements_to_process[requirements_path] = requirements_fp.readlines()
+
+    all_requirements_to_install = {}
+    for requirements_path, pkg_lines in all_requirements_to_process.items():
+        for pkg in pkg_lines:
+            # Remove inline comments which are accepted by pip but not by Home
+            # Assistant's installation method.
+            # https://rosettacode.org/wiki/Strip_comments_from_a_string#Python
+            i = pkg.find("#")
+            if i >= 0:
+                pkg = pkg[:i]
+            pkg = pkg.strip()
+
+            if not pkg:
+                continue
+
+            try:
+                # Attempt to get version of package. Do nothing if it's found since
+                # we want to use the version that's already installed to be safe
+                requirement = Requirement.parse(pkg)
+                pkg_name = requirement.project_name
+
+                # Requirements must be pinned and will be skipped if they aren't
+                if (
+                    len(requirement.specs) != 1
+                    or len(requirement.specs[0]) != 2
+                    or requirement.specs[0][0] != "=="
+                ):
+                    _LOGGER.error(
+                        (
+                            "Ignoring invalid requirement `%s` specified in `%s`, version must "
+                            "be pinned using the format `pkg==version`"
+                        ),
```

Comment:
All the LOG message inner backquotes should normal single quotes; I didn't notice these before.